[Crash] App won't start v0.9.7

Reported from playstore, 6 impacted users

```
java.lang.RuntimeException: Unable to create application im.vector.VectorApp: java.lang.IllegalArgumentException: Invalid URL host: ""
at android.app.ActivityThread.handleBindApplication(ActivityThread.java:5533)
at android.app.ActivityThread.-wrap2(ActivityThread.java)
at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1611)
at android.os.Handler.dispatchMessage(Handler.java:102)
at android.os.Looper.loop(Looper.java:165)
at android.app.ActivityThread.main(ActivityThread.java:6375)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:912)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:802)
Caused by: java.lang.IllegalArgumentException: Invalid URL host: ""
at okhttp3.HttpUrl$Builder.parse(HttpUrl.java:1405)
at okhttp3.HttpUrl.get(HttpUrl.java:916)
at retrofit2.Retrofit$Builder.baseUrl(Retrofit.java:470)
at org.matrix.androidsdk.RestClient.<init>(RestClient.java:169)
at org.matrix.androidsdk.RestClient.<init>(RestClient.java:120)
at org.matrix.androidsdk.rest.client.ThirdPidRestClient.<init>(ThirdPidRestClient.java:60)
at org.matrix.androidsdk.features.identityserver.IdentityServerManager.localSetIdentityServerUrl(IdentityServerManager.kt:278)
at org.matrix.androidsdk.features.identityserver.IdentityServerManager.<init>(IdentityServerManager.kt:69)
at org.matrix.androidsdk.MXSession.<init>(MXSession.java:381)
at org.matrix.androidsdk.MXSession.<init>(MXSession.java:138)
at org.matrix.androidsdk.MXSession$Builder.build(MXSession.java:2753)
at im.vector.Matrix.createSession(Matrix.java:683)
at im.vector.Matrix.createSession(Matrix.java:651)
at im.vector.Matrix.getDefaultSession(Matrix.java:339)
at im.vector.Matrix.addNetworkEventListener(Matrix.java:895)
at im.vector.push.PushManager.<init>(PushManager.java:147)
at im.vector.Matrix.<init>(Matrix.java:207)
at im.vector.Matrix.getInstance(Matrix.java:219)
at im.vector.VectorApp.onCreate(VectorApp.java:235)
at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1025)
at android.app.ActivityThread.handleBindApplication(ActivityThread.java:5530)
```

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -8,7 +8,7 @@ Improvements:
  -
 
 Bugfix:
- -
+ - App won't start with some custom HS config #499
 
 API Change:
  -

--- a/matrix-sdk-crypto/src/main/java/org/matrix/androidsdk/crypto/interfaces/CryptoDataHandler.kt
+++ b/matrix-sdk-crypto/src/main/java/org/matrix/androidsdk/crypto/interfaces/CryptoDataHandler.kt
@@ -17,7 +17,7 @@
 package org.matrix.androidsdk.crypto.interfaces
 
 interface CryptoDataHandler {
-    val store: CryptoStore
+    val store: CryptoStore?
     fun setCryptoEventsListener(eventListener: CryptoEventListener)
     fun getRoom(roomId: String): CryptoRoom
     fun onEventDecrypted(event: CryptoEvent)

--- a/matrix-sdk-crypto/src/main/java/org/matrix/androidsdk/crypto/interfaces/CryptoStore.kt
+++ b/matrix-sdk-crypto/src/main/java/org/matrix/androidsdk/crypto/interfaces/CryptoStore.kt
@@ -18,5 +18,5 @@ package org.matrix.androidsdk.crypto.interfaces
 
 // For classical store
 interface CryptoStore {
-    val eventStreamToken: String
+    val eventStreamToken: String?
 }

--- a/matrix-sdk/src/androidTest/java/org/matrix/androidsdk/crypto/PartialSharedSessionTest.kt
+++ b/matrix-sdk/src/androidTest/java/org/matrix/androidsdk/crypto/PartialSharedSessionTest.kt
@@ -99,8 +99,7 @@ class PartialSharedSessionTest {
         mTestHelper.await(latch)
 
         val aliceRoom = aliceSession.dataHandler
-                .store
-                .getRoom(roomId)
+                .store!!.getRoom(roomId)
 
 
         latch = CountDownLatch(1)
@@ -140,7 +139,7 @@ class PartialSharedSessionTest {
 
         val aliceNewSession = mTestHelper.logIntoAccount(aliceSession.myUserId, defaultSessionParamsWithInitialSync)
 
-        val aliceRoomOtherSession = aliceNewSession.dataHandler.store.getRoom(aliceRoom.roomId)
+        val aliceRoomOtherSession = aliceNewSession.dataHandler.store!!.getRoom(aliceRoom.roomId)
 
         val aliceSecondSessionEvents = sentEvents.map {
             aliceRoomOtherSession.store.getEvent(it.eventId, aliceRoomOtherSession.roomId)

--- a/matrix-sdk/src/main/java/org/matrix/androidsdk/MXDataHandler.java
+++ b/matrix-sdk/src/main/java/org/matrix/androidsdk/MXDataHandler.java
@@ -681,6 +681,7 @@ public class MXDataHandler implements CryptoDataHandler, DataHandlerInterface {
      * @return the used store.
      */
     @Override
+    @Nullable
     public IMXStore getStore() {
         if (isAlive()) {
             return mStore;
@@ -1911,7 +1912,13 @@ public class MXDataHandler implements CryptoDataHandler, DataHandlerInterface {
     public void updateEventState(Event event, Event.SentState newState) {
         if ((null != event) && (event.mSentState != newState)) {
             event.mSentState = newState;
-            getStore().flushRoomEvents(event.roomId);
+            //crash reported on app store
+            if (getStore() != null) {
+                getStore().flushRoomEvents(event.roomId);
+            } else {
+                Log.e(LOG_TAG, "#updateEventState Failed to access to store ");
+            }
+
             onEventSentStateUpdated(event);
         }
     }

--- a/matrix-sdk/src/main/java/org/matrix/androidsdk/MXSession.java
+++ b/matrix-sdk/src/main/java/org/matrix/androidsdk/MXSession.java
@@ -289,11 +289,16 @@ public class MXSession implements CryptoSession {
 
             @Override
             public void onStoreCorrupted(String accountId, String description) {
-                Log.d(LOG_TAG, "## onStoreCorrupted() : token " + getDataHandler().getStore().getEventStreamToken());
+                //NPE reported on store
+                if (getDataHandler().getStore() != null) {
+                    Log.d(LOG_TAG, "## onStoreCorrupted() : token " + getDataHandler().getStore().getEventStreamToken());
 
-                // nothing was saved
-                if (null == getDataHandler().getStore().getEventStreamToken()) {
-                    getDataHandler().onStoreReady();
+                    // nothing was saved
+                    if (null == getDataHandler().getStore().getEventStreamToken()) {
+                        getDataHandler().onStoreReady();
+                    }
+                } else {
+                    Log.d(LOG_TAG, "## onStoreCorrupted() : no store access ");
                 }
             }
 

--- a/matrix-sdk/src/main/java/org/matrix/androidsdk/data/MyUser.java
+++ b/matrix-sdk/src/main/java/org/matrix/androidsdk/data/MyUser.java
@@ -314,7 +314,9 @@ public class MyUser extends User {
                     // local value
                     displayname = aDisplayname;
                     // store metadata
-                    mDataHandler.getStore().setDisplayName(aDisplayname, System.currentTimeMillis());
+                    if (mDataHandler.getStore() != null) {
+                        mDataHandler.getStore().setDisplayName(aDisplayname, System.currentTimeMillis());
+                    }
 
                     mIsDisplayNameRefreshed = true;
 

--- a/matrix-sdk/src/main/java/org/matrix/androidsdk/data/MyUser.java
+++ b/matrix-sdk/src/main/java/org/matrix/androidsdk/data/MyUser.java
@@ -124,17 +124,20 @@ public class MyUser extends User {
      * Build the lists of identifiers
      */
     private void buildIdentifiersLists() {
-        List<ThirdPartyIdentifier> identifiers = mDataHandler.getStore().thirdPartyIdentifiers();
         mEmailIdentifiers = new ArrayList<>();
         mPhoneNumberIdentifiers = new ArrayList<>();
-        for (ThirdPartyIdentifier identifier : identifiers) {
-            switch (identifier.medium) {
-                case ThreePid.MEDIUM_EMAIL:
-                    mEmailIdentifiers.add(identifier);
-                    break;
-                case ThreePid.MEDIUM_MSISDN:
-                    mPhoneNumberIdentifiers.add(identifier);
-                    break;
+        List<ThirdPartyIdentifier> identifiers = mDataHandler.getStore().thirdPartyIdentifiers();
+        //NPE reported on playstore
+        if (mDataHandler.getStore() != null) {
+            for (ThirdPartyIdentifier identifier : identifiers) {
+                switch (identifier.medium) {
+                    case ThreePid.MEDIUM_EMAIL:
+                        mEmailIdentifiers.add(identifier);
+                        break;
+                    case ThreePid.MEDIUM_MSISDN:
+                        mPhoneNumberIdentifiers.add(identifier);
+                        break;
+                }
             }
         }
     }

--- a/matrix-sdk/src/main/java/org/matrix/androidsdk/data/MyUser.java
+++ b/matrix-sdk/src/main/java/org/matrix/androidsdk/data/MyUser.java
@@ -126,9 +126,9 @@ public class MyUser extends User {
     private void buildIdentifiersLists() {
         mEmailIdentifiers = new ArrayList<>();
         mPhoneNumberIdentifiers = new ArrayList<>();
-        List<ThirdPartyIdentifier> identifiers = mDataHandler.getStore().thirdPartyIdentifiers();
         //NPE reported on playstore
         if (mDataHandler.getStore() != null) {
+            List<ThirdPartyIdentifier> identifiers = mDataHandler.getStore().thirdPartyIdentifiers();
             for (ThirdPartyIdentifier identifier : identifiers) {
                 switch (identifier.medium) {
                     case ThreePid.MEDIUM_EMAIL:

--- a/matrix-sdk/src/main/java/org/matrix/androidsdk/features/identityserver/IdentityServerManager.kt
+++ b/matrix-sdk/src/main/java/org/matrix/androidsdk/features/identityserver/IdentityServerManager.kt
@@ -74,7 +74,7 @@ class IdentityServerManager(val mxSession: MXSession,
                 if (accountDataElement.type == AccountDataElement.ACCOUNT_DATA_TYPE_IDENTITY_SERVER) {
                     // The identity server has been updated
                     val accountDataIdentityServer =
-                            mxSession.dataHandler.store.getAccountDataElement(AccountDataElement.ACCOUNT_DATA_TYPE_IDENTITY_SERVER)
+                            mxSession.dataHandler.store?.getAccountDataElement(AccountDataElement.ACCOUNT_DATA_TYPE_IDENTITY_SERVER)
 
                     accountDataIdentityServer?.content?.let {
                         localSetIdentityServerUrl(it[AccountDataElement.ACCOUNT_DATA_KEY_IDENTITY_SERVER_BASE_URL] as String?)
@@ -124,7 +124,7 @@ class IdentityServerManager(val mxSession: MXSession,
      */
     private fun retrieveIdentityServerUrl(): String? {
         val accountDataIdentityServer =
-                mxSession.dataHandler.store.getAccountDataElement(AccountDataElement.ACCOUNT_DATA_TYPE_IDENTITY_SERVER)
+                mxSession.dataHandler.store?.getAccountDataElement(AccountDataElement.ACCOUNT_DATA_TYPE_IDENTITY_SERVER)
 
         accountDataIdentityServer?.content?.let {
             return it[AccountDataElement.ACCOUNT_DATA_KEY_IDENTITY_SERVER_BASE_URL] as String?

--- a/matrix-sdk/src/main/java/org/matrix/androidsdk/features/identityserver/IdentityServerManager.kt
+++ b/matrix-sdk/src/main/java/org/matrix/androidsdk/features/identityserver/IdentityServerManager.kt
@@ -270,12 +270,18 @@ class IdentityServerManager(val mxSession: MXSession,
             identityAuthRestClient = null
             thirdPidRestClient = null
         } else {
-            val alteredHsConfig = HomeServerConnectionConfig.Builder(mxSession.homeServerConfig)
-                    .withIdentityServerUri(Uri.parse(newUrl))
-                    .build()
+            try {
+                val alteredHsConfig = HomeServerConnectionConfig.Builder(mxSession.homeServerConfig)
+                        .withIdentityServerUri(Uri.parse(newUrl))
+                        .build()
 
-            identityAuthRestClient = IdentityAuthRestClient(alteredHsConfig)
-            thirdPidRestClient = ThirdPidRestClient(alteredHsConfig)
+                identityAuthRestClient = IdentityAuthRestClient(alteredHsConfig)
+                thirdPidRestClient = ThirdPidRestClient(alteredHsConfig)
+            } catch (t: Throwable) {
+                Log.e(LOG_TAG, "Failed to create IS Rest clients", t)
+                //What to do from there? this IS is invalid
+                return
+            }
         }
 
         synchronized(listeners) { listeners.forEach { it.onIdentityServerChange() } }
@@ -302,6 +308,11 @@ class IdentityServerManager(val mxSession: MXSession,
     }
 
     private fun requestIdentityServerToken(requestOpenIdTokenResponse: RequestOpenIdTokenResponse, callback: ApiCallback<String>) {
+        if (identityAuthRestClient == null) {
+            callback.onUnexpectedError(IdentityServerNotConfiguredException())
+            return
+        }
+
         identityAuthRestClient?.register(requestOpenIdTokenResponse, object : SimpleApiCallback<IdentityServerRegisterResponse>(callback) {
             override fun onSuccess(info: IdentityServerRegisterResponse) {
                 // Store the token for next time
@@ -596,7 +607,7 @@ class IdentityServerManager(val mxSession: MXSession,
                 override fun onUnexpectedError(e: Exception) {
                     if (e is IdentityServerV2ApiNotAvailable) {
                         //mm ? request to HS?
-                        legacyDeleteAndRequestToken(threePid,callback,idServer)
+                        legacyDeleteAndRequestToken(threePid, callback, idServer)
                     } else {
                         super.onUnexpectedError(e)
                     }
@@ -659,7 +670,7 @@ class IdentityServerManager(val mxSession: MXSession,
                 override fun onUnexpectedError(e: Exception) {
                     if (e is IdentityServerV2ApiNotAvailable) {
                         //mm ? request to HS?
-                        legacyDeleteAndAddMsisdn(threePid,callback,idServer)
+                        legacyDeleteAndAddMsisdn(threePid, callback, idServer)
                     } else {
                         super.onUnexpectedError(e)
                     }

--- a/matrix-sdk/src/main/java/org/matrix/androidsdk/features/terms/TermsManager.kt
+++ b/matrix-sdk/src/main/java/org/matrix/androidsdk/features/terms/TermsManager.kt
@@ -113,7 +113,7 @@ class TermsManager(private val mxSession: MXSession) {
 
     private fun getAlreadyAcceptedTermUrlsFromAccountData(): Set<String> {
         val accountDataCurrentAcceptedTerms =
-                mxSession.dataHandler.store.getAccountDataElement(AccountDataElement.ACCOUNT_DATA_TYPE_ACCEPTED_TERMS)
+                mxSession.dataHandler.store?.getAccountDataElement(AccountDataElement.ACCOUNT_DATA_TYPE_ACCEPTED_TERMS)
 
         return (accountDataCurrentAcceptedTerms?.content
                 ?.get(AccountDataElement.ACCOUNT_DATA_KEY_ACCEPTED_TERMS) as? List<*>)

